<a href="https://colab.research.google.com/github/deliabel/CodeDivisionWorksheets/blob/main/Copy_of_32_How_happy_is_the_world.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# How happy is the world?

![HappinessImage-Benjamin Scott](https://drive.google.com/uc?id=1wWdXTclLAjPpZUiKkR8XGg0Sl7iD1u8T)  

(Image by Benjamin Scott, [source](https://www.natureindex.com/news-blog/data-visualization-these-are-the-happiest-countries-world-happiness-report-twenty-nineteen))   

The Sustainable Development Solutions Network (SDSN) collects data across the world relating to happiness.  They use this data to rank countries in order of happiness factor.

This is not an exact science but can give food for thought in terms of what factors might have the most impact on a nation's happiness levels.

Data is taken from the Gallup World Poll, so not collected directly by SDSN.  
Countries are grouped by region.  

### The factors included are:
Economy (measured in GDP per Capita)
Family (support systems)
Health (measured by Life Expectancy)
Freedom (sense of)
Trust (Government Corruption)
Generosity (charitable inclinations)
Dystopia Residual
*  Dystopic is the theoretical most unhappy country with the lowest levels in all six of the above factors  
*  The Residual measure is a calculated as the average of the six distances from lowest

Let's take a look at the data


---
### Open a data set

Open the data set, an Excel file with only one sheet (so sheet_name is not necessary) from here: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true

Interrogate the data (head, tail, iloc) to get to know what it contains.


In [2]:
import pandas as pd
url_happiness = ' https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true'
happinessdf = pd.read_excel(url_happiness)
happinessdf

Country                           Region  Happiness Rank  \
0    Switzerland                   Western Europe               1   
1        Iceland                   Western Europe               2   
2        Denmark                   Western Europe               3   
3         Norway                   Western Europe               4   
4         Canada                    North America               5   
..           ...                              ...             ...   
153       Rwanda               Sub-Saharan Africa             154   
154        Benin               Sub-Saharan Africa             155   
155        Syria  Middle East and Northern Africa             156   
156      Burundi               Sub-Saharan Africa             157   
157         Togo               Sub-Saharan Africa             158   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
0              7.587         0.03411                   1.39651  1.34951   
1              7.561         0.04884                   1.30232  1.40223   
2              7.527         0.03328                   1.32548  1.36058   
3              7.522         0.03880                   1.45900  1.33095   
4              7.427         0.03553                   1.32629  1.32261   
..               ...             ...                       ...      ...   
153            3.465         0.03464                   0.22208  0.77370   
154            3.340         0.03656                   0.28665  0.35386   
155            3.006         0.05015                   0.66320  0.47489   
156            2.905         0.08658                   0.01530  0.41587   
157            2.839         0.06727                   0.20868  0.13995   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                     0.94143  0.66557                        0.41978   
1                     0.94784  0.62877                        0.14145   
2                     0.87464  0.64938                        0.48357   
3                     0.88521  0.66973                        0.36503   
4                     0.90563  0.63297                        0.32957   
..                        ...      ...                            ...   
153                   0.42864  0.59201                        0.55191   
154                   0.31910  0.48450                        0.08010   
155                   0.72193  0.15684                        0.18906   
156                   0.22396  0.11850                        0.10062   
157                   0.28443  0.36453                        0.10731   

     Generosity  Dystopia Residual  
0       0.29678            2.51738  
1       0.43630            2.70201  
2       0.34139            2.49204  
3       0.34699            2.46531  
4       0.45811            2.45176  
..          ...                ...  
153     0.22628            0.67042  
154     0.18260            1.63328  
155     0.47179            0.32858  
156     0.19727            1.83302  
157     0.16681            1.56726  

[158 rows x 12 columns]

---
### Sort the data in different ways

The data is currently sorted in order of rank.  To sort the data in the table, run the code below, which identifies the column on which to sort in the brackets.

Then, **try sorting on other columns** *Note: you must type the column heading in the quotes and exactly as it appears in the table (including capitalisation)*.  To sort on multiple columns, enter a list of column headings in the brackets (e.g. `.sort_values(['Region','Freedom'])`



In [3]:
sorted_table = happinessdf.sort_values(['Family'])
sorted_table  # output the table below
#                                                                                              ######

Country                      Region  Happiness Rank  \
147  Central African Republic          Sub-Saharan Africa             148   
157                      Togo          Sub-Saharan Africa             158   
152               Afghanistan               Southern Asia             153   
154                     Benin          Sub-Saharan Africa             155   
116                     India               Southern Asia             117   
..                        ...                         ...             ...   
43                 Uzbekistan  Central and Eastern Europe              44   
0                 Switzerland              Western Europe               1   
2                     Denmark              Western Europe               3   
17                    Ireland              Western Europe              18   
1                     Iceland              Western Europe               2   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
147            3.678         0.06112                   0.07850  0.00000   
157            2.839         0.06727                   0.20868  0.13995   
152            3.575         0.03084                   0.31982  0.30285   
154            3.340         0.03656                   0.28665  0.35386   
116            4.565         0.02043                   0.64499  0.38174   
..               ...             ...                       ...      ...   
43             6.003         0.04361                   0.63244  1.34043   
0              7.587         0.03411                   1.39651  1.34951   
2              7.527         0.03328                   1.32548  1.36058   
17             6.940         0.03676                   1.33596  1.36948   
1              7.561         0.04884                   1.30232  1.40223   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
147                   0.06699  0.48879                        0.08289   
157                   0.28443  0.36453                        0.10731   
152                   0.30335  0.23414                        0.09719   
154                   0.31910  0.48450                        0.08010   
116                   0.51529  0.39786                        0.08492   
..                        ...      ...                            ...   
43                    0.59772  0.65821                        0.30826   
0                     0.94143  0.66557                        0.41978   
2                     0.87464  0.64938                        0.48357   
17                    0.89533  0.61777                        0.28703   
1                     0.94784  0.62877                        0.14145   

     Generosity  Dystopia Residual  
147     0.23835            2.72230  
157     0.16681            1.56726  
152     0.36510            1.95210  
154     0.18260            1.63328  
116     0.26475            2.27513  
..          ...                ...  
43      0.22837            2.23741  
0       0.29678            2.51738  
2       0.34139            2.49204  
17      0.45901            1.97570  
1       0.43630            2.70201  

[158 rows x 12 columns]

In [4]:
sorted_table = happinessdf.sort_values(['Happiness Score'])
sorted_table
#                                                   #########

Country                           Region  Happiness Rank  \
157         Togo               Sub-Saharan Africa             158   
156      Burundi               Sub-Saharan Africa             157   
155        Syria  Middle East and Northern Africa             156   
154        Benin               Sub-Saharan Africa             155   
153       Rwanda               Sub-Saharan Africa             154   
..           ...                              ...             ...   
4         Canada                    North America               5   
3         Norway                   Western Europe               4   
2        Denmark                   Western Europe               3   
1        Iceland                   Western Europe               2   
0    Switzerland                   Western Europe               1   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
157            2.839         0.06727                   0.20868  0.13995   
156            2.905         0.08658                   0.01530  0.41587   
155            3.006         0.05015                   0.66320  0.47489   
154            3.340         0.03656                   0.28665  0.35386   
153            3.465         0.03464                   0.22208  0.77370   
..               ...             ...                       ...      ...   
4              7.427         0.03553                   1.32629  1.32261   
3              7.522         0.03880                   1.45900  1.33095   
2              7.527         0.03328                   1.32548  1.36058   
1              7.561         0.04884                   1.30232  1.40223   
0              7.587         0.03411                   1.39651  1.34951   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
157                   0.28443  0.36453                        0.10731   
156                   0.22396  0.11850                        0.10062   
155                   0.72193  0.15684                        0.18906   
154                   0.31910  0.48450                        0.08010   
153                   0.42864  0.59201                        0.55191   
..                        ...      ...                            ...   
4                     0.90563  0.63297                        0.32957   
3                     0.88521  0.66973                        0.36503   
2                     0.87464  0.64938                        0.48357   
1                     0.94784  0.62877                        0.14145   
0                     0.94143  0.66557                        0.41978   

     Generosity  Dystopia Residual  
157     0.16681            1.56726  
156     0.19727            1.83302  
155     0.47179            0.32858  
154     0.18260            1.63328  
153     0.22628            0.67042  
..          ...                ...  
4       0.45811            2.45176  
3       0.34699            2.46531  
2       0.34139            2.49204  
1       0.43630            2.70201  
0       0.29678            2.51738  

[158 rows x 12 columns]

In [5]:
sorted_table = happinessdf.sort_values(['Region','Freedom'])
sorted_table
#                            ########                                                                                     #######

Country                      Region  Happiness Rank  \
8               New Zealand   Australia and New Zealand               9   
9                 Australia   Australia and New Zealand              10   
95   Bosnia and Herzegovina  Central and Eastern Europe              96   
82               Montenegro  Central and Eastern Europe              82   
126                 Armenia  Central and Eastern Europe             127   
..                      ...                         ...             ...   
5                   Finland              Western Europe               6   
2                   Denmark              Western Europe               3   
7                    Sweden              Western Europe               8   
0               Switzerland              Western Europe               1   
3                    Norway              Western Europe               4   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
8              7.286         0.03371                   1.25018  1.31967   
9              7.284         0.04083                   1.33358  1.30923   
95             4.949         0.06913                   0.83223  0.91916   
82             5.192         0.05235                   0.97438  0.90557   
126            4.350         0.04763                   0.76821  0.77711   
..               ...             ...                       ...      ...   
5              7.406         0.03140                   1.29025  1.31826   
2              7.527         0.03328                   1.32548  1.36058   
7              7.364         0.03157                   1.33171  1.28907   
0              7.587         0.03411                   1.39651  1.34951   
3              7.522         0.03880                   1.45900  1.33095   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
8                     0.90837  0.63938                        0.42922   
9                     0.93156  0.65124                        0.35637   
95                    0.79081  0.09245                        0.00227   
82                    0.72521  0.18260                        0.14296   
126                   0.72990  0.19847                        0.03900   
..                        ...      ...                            ...   
5                     0.88911  0.64169                        0.41372   
2                     0.87464  0.64938                        0.48357   
7                     0.91087  0.65980                        0.43844   
0                     0.94143  0.66557                        0.41978   
3                     0.88521  0.66973                        0.36503   

     Generosity  Dystopia Residual  
8       0.47501            2.26425  
9       0.43562            2.26646  
95      0.24808            2.06367  
82      0.16140            2.10017  
126     0.07855            1.75873  
..          ...                ...  
5       0.23351            2.61955  
2       0.34139            2.49204  
7       0.36262            2.37119  
0       0.29678            2.51738  
3       0.34699            2.46531  

[158 rows x 12 columns]

---
### Summarising the data

Look at the happiness dataframe.  Create new dataframes from a range of rows, columns, statistical information, etc.

For each dataframe, add a text cell to explain what it is showing

---
### Next steps

There are data sets for the years 2015 to 2019 available.  To access and try out other years, change 2015 to the required year in the URL in the first code cell.  Leave the rest exactly as it is.  

Other years may have different column headings and so there will be different data to play with.

# stats notes
(7/9/23)

* if normal distribution, use mean. pandas has functions to test for normal distribution. mean is best in general
* median is good if there are lots of outliers
* standard deviation is the avearage distance between each other value and the mean. if it is very wide = unreliable mean. if it is very narow = range is too narrow, eg, if everyone got 50% mark, then the standaer deviation is 0
* it isn't very reliable to say the survivors were mostly of one age if they were all that age. correlations are unreliable if there is a low standard deviation
* anythingmore than 1.5 times then ...(unfinished)
* a way of judging distribution of data set, tells you about spread of data
* the value for standard deviation goes up to the mean (?), and down to 0
   
* similar situation with median and interquartile range.
* if you take the top and bottom quartiles off, and are left with a verly narrow middle range, it indicates the data is not well enough distributed to draw good conlutions. it is low quality data.